nyan/value/number.cpp

```diff
@@ -144,8 +144,8 @@ void Number<T>::apply_value(const Value &value, nyan_op operation) {
 			applier(this->value, change.get(), operation);
 		}
 		else {
-			Float *left = dynamic_cast<Float *>(this);
-			auto change_value = left->handle_infinity(change, operation);
+			Float left = static_cast<Float>(*this);
+			auto change_value = left.handle_infinity(change, operation);
```

Comment:
```
error: invalid ‘static_cast’ from type ‘nyan::Number<long int>*’ to type ‘nyan::Float*’ {aka ‘nyan::Number<double>*’}
[build]   147 |             Float *left = static_cast<Float *>(this);
```

Reply:
Oh, I didn't see that the typecheck above is about operand, not about `this`.

So the error makes sense since `*this` can be `Float`, `Int` or any other template instance of `Number`. If we statically cast `this` to `Float*` in case `this` is actually `Int*`, it would mean we do an invalid side-cast. When we don't have ptrs/refs (like you have now), it actually calls the implicit conversion from `*this` to `Float` (i.e. `Number(T value)` and the to-float/to-int conversion from `operator T()`).

So we do get a new `Float` object that is created from the `int`-object, since it basically calls this: `Float left{this->get()};` But this doesn't handle the cases when the values are infinite. We would need explicit conversions from `Float` to `Int` and vice versa, where we take over the `inf` values.

But what happens when `change_value` (which will be of `Float` here) is `inf`, and we now apply it to `this->value`? It's also not handled, we need to back-convert the float-inf to a int-inf. So i suspect we need a better general approach instead of converting back and forward.

Reply:
(done in #87 )